File: lec01/optional/optional.cpp
// Build with: clang++ -std=c++17 optional.cpp -o optional_cpp

#include <optional>
#include <iostream>

std::optional<std::string> create(bool b) {
	if (b) {
		return std::optional<std::string>{"Hello there!"};
	} else {
		return std::nullopt;
	}
}

int main()
{
	// method one
	auto create_true = create(true);
	if (create_true) {
		std::cout << "create(true) returned "
			      << *create_true
			      << '\n';
	}

	// method two
	std::cout << "create(false) returned "
		      << create(false).value_or("<empty>")
		      << '\n';
}

File: lec03/optional/optional.go
// run with `go run optional.go`

package main

import "fmt"

func strPtr(s string) *string {
	return &s
}

func create(b bool) *string {
	if (b) {
		return strPtr("Hello there")
	} else {
		return nil
	}
}

func main() {
	// method 1
	create_true := create(true)
	if (create_true != nil) {
		fmt.Printf("create(true) returned %s\n", *create_true)
	}

	// method 2
	create_false := create(false)
	var string_to_print string

	if (create_false != nil) {
		string_to_print = *create_false
	} else {
		string_to_print = "<empty>"
	}

	fmt.Printf("create(false) returned %s\n", string_to_print)
}
